external-crates/move/crates/move-compiler/src/typing/core.rs

```diff
@@ -194,20 +248,134 @@ impl<'env> Context<'env> {
         tn: &TypeName,
         method: Name,
     ) -> Option<(ModuleIdent, FunctionName)> {
+        let cur_color = self.use_funs.last().unwrap().color;
         self.use_funs.iter_mut().rev().find_map(|scope| {
+            // scope color is None for global scope, which is always in consideration
+            // otherwise, the color must match the current color. In practice, we are preventing
+            // macro scopes from interfering with each the scopes in which they are expanded
+            if scope.color.is_some() && scope.color != cur_color {
+                return None;
+            }
             let use_fun = scope.use_funs.get_mut(tn)?.get_mut(&method)?;
             use_fun.used = true;
             Some(use_fun.target_function)
         })
     }
 
+    /// true iff it is safe to expand,
+    /// false with an error otherwise (e.g. a recursive expansion)
+    pub fn add_macro_expansion(&mut self, m: ModuleIdent, f: FunctionName, loc: Loc) -> bool {
+        let current_call_color = self.use_funs.last().unwrap().color.unwrap();
+
+        let mut prev_opt = None;
+        for (idx, mexp) in self.macro_expansion.iter().enumerate().rev() {
+            match mexp {
+                MacroExpansion::Lambda { use_funs_scope } => {
+                    // the lambda has a smaller (or equal) color, meaning this lambda was written
+                    // and invoked from an outer scope
+                    if current_call_color > *use_funs_scope {
+                        break;
+                    }
+                }
+                MacroExpansion::Call {
+                    module,
+                    function,
+                    use_funs_scope,
+                    ..
+                } => {
+                    // recursive call is found if we call the same module
+                    // but at a color that is less than or equal to the current color is safe
+                    // since it means it is from some outer scope (likely from a lambda)
+                    if current_call_color > *use_funs_scope && module == &m && function == &f {
+                        prev_opt = Some(idx);
+                        break;
+                    }
+                }
+            }
+        }
+
+        if let Some(idx) = prev_opt {
+            let msg = format!(
+                "Recursive macro expansion. '{}::{}' has already been expanded",
```

Comment:
```suggestion
                "Recursive macro expansion. '{}::{}' cannot re-expand to itself during expansion",
```

Could even consider printing the expansion "stack trace" here!